presto-main/src/main/java/io/prestosql/sql/planner/iterative/IterativeOptimizer.java

```diff
@@ -182,15 +188,25 @@ private boolean exploreNode(int group, Context context)
         return Rule.Result.empty();
     }
 
-    private boolean exploreChildren(int group, Context context)
+    private boolean exploreChildren(int group, Context context, Context deleteContext)
     {
         boolean progress = false;
 
         PlanNode expression = context.memo.getNode(group);
-        for (PlanNode child : expression.getSources()) {
+        for (int i = 0; i < expression.getSources().size(); i++) {
+            PlanNode child = expression.getSources().get(i);
             checkState(child instanceof GroupReference, "Expected child to be a group reference. Found: " + child.getClass().getName());
 
-            if (exploreGroup(((GroupReference) child).getGroupId(), context)) {
+            Context childContext;
+            if (i == 0) {
+                // pass the context of Delete to the left branch of plan
```

Comment:
Maybe more elegant way would be to add `boolean forDelete` to `TableScanNode`. Then we could simply use `PlanNodeSearcher` within a rule to search for `TableScans` with such flag.

In current approach `delete` logic (context goes to left) becomes part of `IterativeOptimizers` itself